### PR TITLE
#2758: Implement fail-fast logic for dialect resolution

### PR DESCRIPTION
**Pull request by kibertoad**
refs https://github.com/tgriesser/knex/issues/2758

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -1,73 +1,3 @@
-import Raw from './raw';
-import Client from './client';
+const Knex = require('./knex');
```
Comment (kibertoad) [inline src/index.js]: Rationale behind this change: it it very counter-intuitive to write tests for a file named "index.js" which is in general supposed to be a catalogue and do not contain logic on its own.

File: src/knex.js [hunk 2]
```diff
@@ -0,0 +1,83 @@
+import Raw from './raw';
+import Client from './client';
+
+import makeKnex from './util/make-knex';
+import parseConnection from './util/parse-connection';
+import { SUPPORTED_CLIENTS, CLIENT_ALIASES } from './constants';
+
+export default function Knex(config) {
+  // If config is string, try to parse it
```
Comment (kibertoad) [inline src/knex.js]: added comment with explanation for each block, please verify the accuracy of them

File: src/knex.js [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+  if (typeof config === 'string') {
+    const parsedConfig = Object.assign(parseConnection(config), arguments[2]);
```
Comment (kibertoad) [inline src/knex.js]: _.assign replaced with Object.assign in this file since it is supported since Node 4.

File: src/knex.js [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+/* eslint no-console:0 */
+
+Object.defineProperties(Knex, {
+  VERSION: {
+    get() {
+      console.warn(
+        'Knex.VERSION is deprecated, you can get the module version' +
+          "by running require('knex/package').version"
+      );
+      return '0.12.6';
```
Comment (ricardograca) [inline src/knex.js]: Looks like this return value hasn't been updated in some time :laughing: Maybe it's time to remove this deprecation warning? I know it's not related to the rest of the work done here, so feel free to ignore this comment.

Comment (kibertoad) [reply src/knex.js]: So you suggest to remove entire
```
VERSION: {
    get() {
      console.warn(
        'Knex.VERSION is deprecated, you can get the module version' +
          "by running require('knex/package').version"
      );
      return '0.12.6';
}
```

block? Sure, I can do that.

Comment (ricardograca) [reply src/knex.js]: Yes. The message has been there for almost 2 years now. I think that's enough time for a deprecation warning.

**Comment by kibertoad** [comment]
@ricardograca Did the change, anything else?..

**Comment by ricardograca** [comment]
Cool! I'll take a look at the rest of changes later today.

File: src/constants.js [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+  postgresql: 'postgres',
+  sqlite: 'sqlite3',
+});
+
+const SUPPORTED_CLIENTS = Object.freeze(
+  [
+    'mssql',
+    'mysql',
+    'mysql2',
+    'oracle',
```
Comment (elhigu) [inline src/constants.js]: I  think that oracle is not actually supported by knex. It is just used by oracledb, which is supported.

File: src/client.js [hunk 6]
```diff
@@ -38,7 +39,8 @@ function Client(config = {}) {
   //Client is a required field, so throw error if it's not supplied.
   //If 'this.dialect' is set, then this is a 'super()' call, in which case
   //'client' does not have to be set as it's already assigned on the client prototype.
-  if (!this.config.client && !this.dialect) {
+  const dbClient = this.config.client || this.dialect;
```
[the remark this answers is not in the document]
Comment (kibertoad) [reply src/client.js]: Would you like me to log a deprecation warning when this.dialect is set?

File: src/constants.js (hunk 5 above)
Comment (kibertoad) [reply src/constants.js]: Thanks, will remove

Comment (elhigu) [reply src/constants.js]: afaik. it is not tested either

File: src/index.js [hunk 7]
```diff
@@ -1,73 +1,3 @@
...
-/* eslint no-console:0 */
-
-Object.defineProperties(Knex, {
-  VERSION: {
-    get() {
-      console.warn(
-        'Knex.VERSION is deprecated, you can get the module version' +
-          "by running require('knex/package').version"
-      );
-      return '0.12.6';
```
Comment (elhigu) [inline src/index.js]: Was this just dropped in this PR? It's hard to follow all the changes after renaming... I need to diff the files locally. If this was dropped, it must be notified in changelog as breaking change.

Comment (kibertoad) [reply src/index.js]: I've added comments in the PR for all the changes after renaming; this one was done after @ricardograca comment. If you are ok with this change, I'll add it to changelog.

Comment (elhigu) [reply src/index.js]: Yep, I'm fine with it. I'll check out diff locally later on.

File: src/client.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (elhigu) [reply src/client.js]: That is a good idea, but we should discuss about it with other contributors too if they agree that alias is not wanted. My opinion is based on some previous problems with testing all these aliases and many times tests has been depending on .client and then they have failed when .dialect has been used by some other driver setup...

[the remark this answers is not in the document]
Comment (elhigu) [reply src/client.js]: @tgriesser @wubzz maybe has an opinion

[the remark this answers is not in the document]
Comment (wubzz) [reply src/client.js]: I think it's a good idea to remove "dialect" so long as the knex internals are also adjusted to reflect this. For example `src/dialects` -> `src/clients`.

[the remark this answers is not in the document]
Comment (kibertoad) [reply src/client.js]: Considering that there was no deprecation period with warnings prior to that, should we really just drop it immediately?

**Comment by kibertoad** [comment]
@elhigu Changes addressed.

**Comment by elhigu** [comment]
Looks like tests needs still some fixing so that they stop using .dialect too. Now at least custom logger tests are failing because of unexpected deprecation warnings.

File: CHANGELOG.md [hunk 8]
```diff
@@ -4,6 +4,7 @@
 ### Breaking Changes:
 
 - Use datetime2 for MSSQL datetime + timestamp types. This change is incompatible with MSSQL older than 2008 #2757
+- Knex.VERSION() method was removed, run "require('knex/package').version" instead.
```
Comment (elhigu) [inline CHANGELOG.md]: Please add also #2776 in the end of line so that this PR is referenced

**Comment by kibertoad** [comment]
@elhigu Fixed!

**Comment by kibertoad** [comment]
Aaaand it's green again.

**Comment by kibertoad** [comment]
@elhigu Is anything currently missing?

**Comment by elhigu** [comment]
Looks good to me, thanks!
